core-bundle/contao/drivers/DC_Table.php

```diff
@@ -4326,11 +4326,25 @@ protected function generateTree($table, $id, $arrPrevNext, $blnHasSorting, $intM
 			$currentRecord = null;
 		}
 
+		// Special handling for visible root trails, which are tree nodes the user does not have access to.
+		// $this->getCurrentRecord() can deny access, but we still need the record to render a label.
+		if (null === $currentRecord && !empty($this->visibleRootTrails))
+		{
+			$currentRecord = Database::getInstance()
+				->prepare("SELECT * FROM " . $table . " WHERE id=?")
+				->execute($id)
+				->fetchAssoc()
+			;
+
+			if (!$currentRecord || $checkIdAllowed ? !\in_array($id, $this->visibleRootTrails) : !\in_array($currentRecord['pid'] ?? null, $this->visibleRootTrails))
+			{
+				return '';
+			}
+		}
+
 		// Return if there is no result
 		if (null === $currentRecord)
 		{
-			$objSessionBag->replace($session);
```

Comment:
Is this a related change?

Reply:
it is only related in the sense that I noticed it is completely useless (and therefore I could also do an early-return above). 😆